### ISSUE TITLE
Add Falcon security mode to nvbios

### ISSUE THREAD

**Issue by Panovich** [envytools/envytools#200]
It would be useful

**Comment by Echelon9** [comment]
Can you explain in more detail what you mean in this request? (I appreciate English may not be your native language).
The security modes of individual Falcon chips on a graphics card are "stateful" at runtime, e.g. they can be loaded into High Security mode, or Low Security mode, or No Security mode. In each state, different registers and functionality are enabled.
However, whilst we can infer from a VBIOS which GPU family is present and therefore which Falcon chips are likely on the board, it doesn't mandate security levels in the VBIOS.
Falcon security levels are a hardware feature.

**Comment by Panovich** [comment]
> The security modes of individual Falcon chips on a graphics card are "stateful" at runtime, e.g. they can be loaded into High Security mode, or Low Security mode, or No Security mode. In each state, different registers and functionality are enabled.

On GM206+, the PMU on most firmwares does not allow any changes to devinit. It needs a signature. But there are firmware versions that allow changes. I realized it too late, and I didn't keep it. I looked at the MODS Report, and on one of the firmwares, the modified devinit was executed by the PMU. From this, I concluded that the security level is set somewhere during the initialization of the video chip. But I can't find out where, I don't even have anything to compare it to. On download.nvidia.com it is written that falcon security also protects devinit. Need to find out where this protection is enabled.

P.S. I thought that this might be strap config, but I didn't find anything. There are only useless settings)

**Comment by Panovich** [comment]
I think I noticed the same firmware on GP107, but I can't find out, because I don't know where to get MODS newer than version 367, I'm not sure, since I couldn't run any GP107 properly, but the desktop version reacts to any devinit change with a BSOD (protection is probably enabled), and on the laptop version just code 43 :)
GM206 - always BSOD, same reason, and MODS reports
NVRM: DevinitPmuOffloadDevinitToPmu Devinit complete is false
NVRM: bp @ ../../../../resman/kernel/devinit/nv/devinit_pmu.c:384

**Comment by Panovich** [comment]
There is a pointer to init script in PMU DMEM Section

**Comment by Echelon9** [comment]
> On GM206+, the PMU on most firmwares does not allow any changes to devinit. It needs a signature.

Yes, nouveau developers know this.

> But there are firmware versions that allow changes.

Do you have evidence of this? Without better understanding these firmwares, we're not going to make much progress. I suspect, if they do exist, that they are non-redistributable internal NVIDIA engineering firmwares.

> On download.nvidia.com it is written that falcon security also protects devinit.

DEVINIT is a simple scripting language which allows GPU programming sequences to be shared by multiple entities that would otherwise not be able to execute common code.

For example, a single devinit script could be run by the PMU Falcon, *"the VGA BIOS, the Resource Manager [this is NVIDIA's name for their binary kernel driver] for secondary adapter cases, and the FCode firmware for the Apple Mac platforms."* The package of changes to the GPU state should be the same across all if their devinit interpreters work per the documented specifications.
See: https://nvidia.github.io/open-gpu-doc/Devinit/devinit.xml

So what does the Falcon security model protect? Open source driver developers care, because opcodes within the devinit scripts can program useful GPU registers that are only available from Falcon chips (or other DEVINIT interpreter) when in High Security mode. We understand this to be an on-silicon hardware restriction.

No boot Falcon in HS mode, which requires scripts to have been NVIDIA signed == no access to certain useful GPU registers.

What you are describing is a fundamental break in the implementation or cryptographic system that NVIDIA used to enforce those controls.

**Comment by Panovich** [comment]
> Do you have evidence of this? Without better understanding these firmwares, we're not going to make much progress. I suspect, if they do exist, that they are non-redistributable internal NVIDIA engineering firmwares.
Wow, if this is so serious, I'll try to find it again. I spent a long time picking up from different video cards on the same chip, from different laptops. I think I suspect where I got this firmware then, maybe I will find it again) If I then knew how difficult it is, I would have saved it. I only realized this now when I needed to change devinit. Before that, I was on GM107, but there was no HS mode) There seems to be no protection at all. Therefore, I did not think that on newer chips this may cause a problem) Now it does not even allow DCB to change
Do you know which registers it doesn't allow to change in NS mode? And can I run devinit without a signature? Maybe I'll just delete these registers from there, my video card is initialized as a 3d accelerator, it doesn't even have video outputs.
It looks like falcon has debug mode. Could it be that it was enabled on the firmware that allowed the changes? Or is it turned on by some kind of hardware strap?
Thank you for your answer, I'll go look for that firmware.

**Comment by Echelon9** [comment]
> Do you know which registers it doesn't allow to change in NS mode?

Relevant ones include fan controls (e.g. no point controlling higher performance clock speeds if the card can't regulate it's own cooling system). This isn't always an issue with mobile GPUs, as the fans can be controlled from the main CPU embedded controller, but on any discrete card doing performance reclocking without fan control is a quick way to kill your card.

> It looks like falcon has debug mode. Could it be that it was enabled on the firmware that allowed the changes? Or is it turned on by some kind of hardware strap?

Falcon debug registers provide access to the "instruction pointer", other registers and ability to set breakpoints on the small MCUs which are the various Falcon chips on the GPU card. It is enabled by default.

We already have basic working knowledge of it. I believe @karolherbst had a simple bash script that provided simple GDB-like functionality. But again, I recall these debug registers may be partially locked down once the Falcon enters HS mode.

**Comment by Panovich** [comment]
Yes, I can remove all thermal management, in the laptop it is not necessary. Videobios doesn't even have a Fan Control table.
So, can I run devinit somehow? Without a signature. The PMU doesn't want to do this in any way. I need to use other gpio table and 8gb 7000mhz. Any change in DCB or in any of the scripts, and video card throws: Devinit complete is false. 
By the way, the firmware from gtx on 750 (GM206) is very strange. When trying to run: Primary Devinit on PMU is not available!. Envytools also does not find the falcon ucode table. But physically it is present. I'll check it out. Maybe this is a chance to catch on. pmu bootloader looks a bit different.

**Comment by Panovich** [comment]
To correctly recognize the falcon ucode table on some video cards, you need to make some changes. 
1) Some video cards have this table at an address greater than 16 bits. Need to change: 
nvbios/p.c - struct p_known_tables  - uint32_t *ptr;
nvbios/bios.h - struct envy_bios_p_falcon_ucode and struct envy_bios_p_falcon_ucode_entry - uint32_t offset;
2) GM107 does not have IMEM sec. I suggest this code: 
nvbios/p.c - in void envy_bios_print_p_falcon_ucode - 
`			if (desc->stored_size == desc->uncompressed_size) {
				if (mask & ENVY_BIOS_PRINT_VERBOSE) fprintf(out, "Falcon UCODE %i: bootloader (IMEM):\n", i);
				if (desc->imem_sec_base == 0) {
					envy_bios_dump_hex(bios, out, desc->offset + 0x30, desc->dmem_offset /*desc->imem_load_size - desc->imem_sec_size*/, mask);
					if (mask & ENVY_BIOS_PRINT_VERBOSE) fprintf(out, "Falcon UCODE %i: ucode (IMEM sec) NOT PRESENT\n", i);
				} else {
					envy_bios_dump_hex(bios, out, desc->offset + 0x30, desc->imem_sec_base /*desc->imem_load_size - desc->imem_sec_size*/, mask);
					if (mask & ENVY_BIOS_PRINT_VERBOSE) fprintf(out, "Falcon UCODE %i: ucode (IMEM sec):\n", i);
					envy_bios_dump_hex(bios, out, desc->offset + 0x30 + desc->imem_sec_base /*desc->imem_load_size - desc->imem_sec_size*/, desc->imem_sec_size, mask);
				}
				if (mask & ENVY_BIOS_PRINT_VERBOSE) fprintf(out, "Falcon UCODE %i: data (DMEM):\n", i);
				envy_bios_dump_hex(bios, out, desc->offset + 0x30 + desc->dmem_offset, desc->dmem_load_size, mask);
			}`

**Comment by imirkin** [comment]
GM107's have the beginnings of secure modes that were made required in GM20x:

https://nvidia.github.io/open-gpu-doc/Falcon-Security/Falcon-Security.html

Also note that the VBIOS may have multiple sections. This is not implemented in nvbios, but nouveau does implement it:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/gpu/drm/nouveau/nvkm/subdev/bios/shadow.c

The code's a bit hard to follow, but shadow_image() calls itself recursively until it hits the last image. I don't remember offhand what the "later" images are used for, but clearly something since the functionality was added.

**Comment by Panovich** [comment]
> GM107's have the beginnings of secure modes that were made required in GM20x:

Yes, i know. Thank you. I read the entire open gpu doc.
I'm looking for how to get the pmu to run a modified devinit on gm206 and later
What I presented above is necessary in order for nvbios to correctly parse gm107 firmware.
And need to do a code for those firmware versions where part 0 and part 1 are located next to each other. I also wrote about this above. The remaining VN sections are located after these parts, and to get to the falcon, you need to skip part 1
